core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -378,55 +381,157 @@ public Object visitGetAllCommand(InvocationContext ctx, GetAllCommand command) t
-                     CacheEntry entry = value == null ? NullCacheEntry.getInstance() : value.toInternalCacheEntry(key);
-                     wrapRemoteEntry(ctx, key, entry, false);
-                  }
-                  counterValue = --allFuture.counter;
+            return;
+         }
+         Object responseValue = ((SuccessfulResponse) response).getResponseValue();
+         if (responseValue instanceof InternalCacheValue[]) {
+            InternalCacheValue[] values = (InternalCacheValue[]) responseValue;
+            synchronized (allFuture) {
+               // Check if other handlers haven't finished with an exception
+               if (allFuture.isDone()) {
+                  return;
                }
-               // complete the future after sync block!
-               if (counterValue == 0) {
+               for (int i = 0; i < keys.size(); ++i) {
+                  Object key = keys.get(i);
+                  InternalCacheValue value = values[i];
+                  CacheEntry entry = value == null ? NullCacheEntry.getInstance() : value.toInternalCacheEntry(key);
+                  wrapRemoteEntry(ctx, key, entry, false);
+               }
+               if (--allFuture.counter == 0) {
                   allFuture.complete(null);
                }
-            } else {
-               allFuture.completeExceptionally(new IllegalStateException("Unexpected response value: " + responseValue));
             }
-         });
+         } else {
+            allFuture.completeExceptionally(new IllegalStateException("Unexpected response value: " + responseValue));
+         }
+      }
+
+      private void handleMissingResponse() {
+         GlobalTransaction gtx = ctx.isInTxScope() ? ((TxInvocationContext) ctx).getGlobalTransaction() : null;
+
+         LocalizedCacheTopology cacheTopology = checkTopologyId(command);
+         Map<Address, List<Object>> requestedKeys = getKeysByOwner(ctx, keys, cacheTopology, null, contactedNodes);
+         synchronized (allFuture) {
+            allFuture.counter += requestedKeys.size();
+         }
+         for (Map.Entry<Address, List<Object>> pair : requestedKeys.entrySet()) {
+            ClusteredGetAllCommand clusteredGetAllCommand = cf.buildClusteredGetAllCommand(pair.getValue(), command.getFlagsBitSet(), gtx);
+            clusteredGetAllCommand.setTopologyId(command.getTopologyId());
+            List<Address> contactedNodes = new ArrayList<>(this.contactedNodes.size() + 1);
+            contactedNodes.addAll(this.contactedNodes);
+            contactedNodes.add(pair.getKey());
+            keys.removeAll(pair.getValue());
+            rpcManager.invokeRemotelyAsync(Collections.singleton(pair.getKey()), clusteredGetAllCommand, syncIgnoreLeavers)
+                  .whenComplete(new ClusteredGetAllHandler(allFuture, ctx, command, pair.getValue(), contactedNodes));
+         }
+         if (!keys.isEmpty()) {
+            // current keys don't have any owner left, we have to remove them
```

Comment:
When the response for the clustered get, the entries are wrapped into context. The GetAllCommand.perform then reads those entries and puts the values into the map. If we don't get any response, it's not wrapped, and it's illegal to not find an entry in context when the command is executed.